pkg/streamingpromql/operators/aggregations/quantile.go

```diff
@@ -0,0 +1,180 @@
+	Aggregation              *Aggregation
+	MemoryConsumptionTracker *limiting.MemoryConsumptionTracker
+	Annotations              *annotations.Annotations
+}
+
+func NewQuantileAggregation(
+	inner types.InstantVectorOperator,
+	param types.ScalarOperator,
+	timeRange types.QueryTimeRange,
+	grouping []string,
+	without bool,
+	memoryConsumptionTracker *limiting.MemoryConsumptionTracker,
+	annotations *annotations.Annotations,
+	expressionPosition posrange.PositionRange,
+) (*QuantileAggregation, error) {
+
+	a, err := NewAggregation(
+		inner,
+		timeRange,
+		grouping,
+		without,
+		parser.QUANTILE,
+		memoryConsumptionTracker,
+		annotations,
+		expressionPosition,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	q := &QuantileAggregation{
+		Aggregation:              a,
+		Param:                    param,
+		MemoryConsumptionTracker: memoryConsumptionTracker,
+		Annotations:              annotations,
+	}
+
+	return q, nil
+}
+
+func (q *QuantileAggregation) ExpressionPosition() posrange.PositionRange {
+	return q.Aggregation.ExpressionPosition()
+}
+
+func (q *QuantileAggregation) SeriesMetadata(ctx context.Context) ([]types.SeriesMetadata, error) {
+	var err error
+	q.Aggregation.ParamData, err = q.Param.GetValues(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// Validate the parameter now so we only have to do it once for each group
+	for _, p := range q.Aggregation.ParamData.Samples {
+		if math.IsNaN(p.F) || p.F < 0 || p.F > 1 {
+			q.Annotations.Add(annotations.NewInvalidQuantileWarning(p.F, q.Param.ExpressionPosition()))
+		}
+	}
+
+	return q.Aggregation.SeriesMetadata(ctx)
+}
+
+func (q *QuantileAggregation) NextSeries(ctx context.Context) (types.InstantVectorSeriesData, error) {
+	return q.Aggregation.NextSeries(ctx)
+}
+
+func (q *QuantileAggregation) Close() {
+	if q.Aggregation.ParamData.Samples != nil {
+		types.FPointSlicePool.Put(q.Aggregation.ParamData.Samples, q.MemoryConsumptionTracker)
+	}
+	if q.Param != nil {
+		q.Param.Close()
+	}
+	q.Aggregation.Close()
+}
+
+type QuantileAggregationGroup struct {
+	qGroups []qGroup // A group per point in time
+}
+
+type qGroup struct {
+	points []float64 // All of the floats for this group of series at a point in time
+}
+
+const maxExpectedQuantileGroups = 64 // There isn't much science to this
+
+var qGroupPool = types.NewLimitingBucketedPool(
+	pool.NewBucketedPool(maxExpectedQuantileGroups, func(size int) []qGroup {
+		return make([]qGroup, 0, size)
+	}),
+	uint64(unsafe.Sizeof(qGroup{})),
+	false,
+	nil,
+)
+
+func (q *QuantileAggregationGroup) AccumulateSeries(data types.InstantVectorSeriesData, timeRange types.QueryTimeRange, memoryConsumptionTracker *limiting.MemoryConsumptionTracker, emitAnnotationFunc types.EmitAnnotationFunc, remainingSeriesInGroup uint) error {
+	defer types.PutInstantVectorSeriesData(data, memoryConsumptionTracker)
+
+	if len(data.Histograms) > 0 {
+		emitAnnotationFunc(func(_ string, expressionPosition posrange.PositionRange) error {
```

Comment:
[nit] I realise this was what it was called previously, but I think `emitAnnotation` would be a better name for this parameter, with `EmitAnnotationFunc` remaining as the type.

(could be something for a follow-up PR)

Reply:
Will leave for followup

Reply:
Done in https://github.com/grafana/mimir/pull/10885